apiml/src/main/java/org/zowe/apiml/filter/CategorizeCertsWebFilter.java

```diff
@@ -0,0 +1,200 @@
+    /**
+     * Categorizes certificates from the TLS handshake and/or a specific header.
+     * Stores categorized certificates in the exchange attributes.
+     *
+     * @param exchange The current server web exchange.
+     */
+    private void categorizeCerts(ServerWebExchange exchange) {
+        Optional<X509Certificate[]> certsFromTlsOpt = Optional.of(exchange)
+            .map(ServerWebExchange::getRequest)
+            .map(ServerHttpRequest::getSslInfo)
+            .map(SslInfo::getPeerCertificates)
+            .filter(Objects::nonNull)
+            .filter(ssl -> ssl.length > 0);
+
+        certsFromTlsOpt.ifPresent(certsFromTls -> {
+            Optional<X509Certificate> clientCertFromHeader = getClientCertFromHeader(exchange.getRequest());
+            ServerHttpRequest.Builder requestBuilder = exchange.getRequest().mutate();
+
+            if (certificateValidator.isForwardingEnabled() &&
+                certificateValidator.hasGatewayChain(certsFromTls) &&
+                clientCertFromHeader.isPresent()) {
+
+                certificateValidator.updateAPIMLPublicKeyCertificates(certsFromTls);
+
+                X509Certificate[] clientAuthCerts = selectCerts(
+                    new X509Certificate[]{clientCertFromHeader.get()},
+                    certificateForClientAuth
+                );
+                exchange.getAttributes().put(ATTR_NAME_CLIENT_AUTH_X509_CERTIFICATE, clientAuthCerts);
+                log.debug(LOG_FORMAT_FILTERING_CERTIFICATES, ATTR_NAME_CLIENT_AUTH_X509_CERTIFICATE, Arrays.toString(clientAuthCerts));
+
+                exchange.getAttributes().put(ATTR_NAME_JAKARTA_SERVLET_REQUEST_X509_CERTIFICATE, certsFromTls);
+                log.debug("Retaining full TLS certificate chain in attribute {}: {}", ATTR_NAME_JAKARTA_SERVLET_REQUEST_X509_CERTIFICATE, Arrays.toString(certsFromTls));
+
+                var sslInfo = SimpleSslInfo.builder().peerCertificates(clientAuthCerts).build();
+                requestBuilder.sslInfo(sslInfo);
+
+            } else {
+                X509Certificate[] clientAuthCerts = selectCerts(certsFromTls, certificateForClientAuth);
+                exchange.getAttributes().put(ATTR_NAME_CLIENT_AUTH_X509_CERTIFICATE, clientAuthCerts);
+                log.debug(LOG_FORMAT_FILTERING_CERTIFICATES, ATTR_NAME_CLIENT_AUTH_X509_CERTIFICATE, Arrays.toString(clientAuthCerts));
+
+                X509Certificate[] apimlFilteredCerts = selectCerts(certsFromTls, apimlCertificate);
+                exchange.getAttributes().put(ATTR_NAME_JAKARTA_SERVLET_REQUEST_X509_CERTIFICATE, apimlFilteredCerts);
+                log.debug(LOG_FORMAT_FILTERING_CERTIFICATES, ATTR_NAME_JAKARTA_SERVLET_REQUEST_X509_CERTIFICATE, Arrays.toString(apimlFilteredCerts));
+                var sslInfo = SimpleSslInfo.builder().peerCertificates(clientAuthCerts).build();
+                requestBuilder.sslInfo(sslInfo);
+
+            }
+
+            exchange.mutate().request(requestBuilder.build()).build();
```

Comment:
Shouldn't we return the new instance of Exchange? I guess in this way it is not certain that SslInfo is update (see the previous instance).